File: go/videocall/service.go
// SPDX-License-Identifier: MIT

// Package videocall implements video call-control over PacketType.VideoCall
// (PacketType 27) for the Aether mesh — the caller-intent signalling layer
// (ring / accept / decline / hangup) directed between two peers, distinct from
// the media plane (SDP/ICE negotiation + frames) handled separately by the
// streaming VideoCall service. The caller rings a peer (minting a call id); either
// side then accepts, declines, or hangs up. Each control signal is a directed
// VideoCall packet sent to the peer as the next hop; inbound signals surface via
// OnCallStateChanged. Mirrors the C# AetherNet.VideoCallControl.VideoCallControlService.
package videocall

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/bhengubv/aether-protocol/go/constants"
	"github.com/bhengubv/aether-protocol/go/protocol"
	"github.com/bhengubv/aether-protocol/go/routing"
)

// CallStateChanged is surfaced when a video call-control signal arrives from a
// peer. Mirrors the C# VideoCallStateChanged event args.
type CallStateChanged struct {
	// CallId is the id of the call the signal refers to.
	CallId uuid.UUID
	// Action is the control verb received ("ring" / "accept" / "decline" / "hangup").
	Action string
	// FromUhid is the UHID of the peer that sent the signal (the packet source).
	FromUhid string
}

// Service is the default video call-control service. It sends directed VideoCall
// signals (ring/accept/decline/hangup) and surfaces inbound ones via
// OnCallStateChanged.
type Service struct {
	sender routing.MeshSender

	// OnCallStateChanged fires when a call-control signal is received from a peer.
	// Mirrors the C# CallStateChanged event.
	OnCallStateChanged func(evt CallStateChanged)
}

// NewService constructs a Service. Panics if sender is nil.
func NewService(sender routing.MeshSender) *Service {
	if sender == nil {
		panic("videocall: sender must not be nil")
	}
	return &Service{sender: sender}
}

// Ring rings peerUhid: mints a call id and sends a directed "ring". Returns the
// new call id (and any send error). Returns an error if peerUhid is empty.
func (s *Service) Ring(ctx context.Context, peerUhid string) (uuid.UUID, error) {
	if peerUhid == "" {
		return uuid.Nil, errors.New("videocall: peerUhid must not be empty")
	}
	callID := uuid.New()
	if _, err := s.sendControl(ctx, callID, peerUhid, "ring"); err != nil {
		return uuid.Nil, err
	}
	return callID, nil
}

// Accept sends a directed "accept" for callId to peerUhid. Returns delivery success.
func (s *Service) Accept(ctx context.Context, callID uuid.UUID, peerUhid string) (bool, error) {
	return s.sendControl(ctx, callID, peerUhid, "accept")
}

// Decline sends a directed "decline" for callId to peerUhid. Returns delivery success.
func (s *Service) Decline(ctx context.Context, callID uuid.UUID, peerUhid string) (bool, error) {
	return s.sendControl(ctx, callID, peerUhid, "decline")
}

// Hangup sends a directed "hangup" for callId to peerUhid. Returns delivery success.
func (s *Service) Hangup(ctx context.Context, callID uuid.UUID, peerUhid string) (bool, error) {
	return s.sendControl(ctx, callID, peerUhid, "hangup")
}

// sendControl builds and sends a directed VideoCall control packet (dest=peer,
// ttl=constants.DefaultTtl) carrying {call_id, action, sent_at_ms:now}.
func (s *Service) sendControl(ctx context.Context, callID uuid.UUID, peerUhid, action string) (bool, error) {
	if peerUhid == "" {
		return false, errors.New("videocall: peerUhid must not be empty")
	}

	body, err := json.Marshal(videoCallControlWire{
		CallID:   callID,
		Action:   action,
		SentAtMs: time.Now().UnixMilli(),
	})
	if err != nil {
		return false, fmt.Errorf("videocall: marshal payload: %w", err)
	}

	pkt := protocol.NewMeshPacket()
	pkt.Type = protocol.VideoCallPkt
	pkt.SourceUhid = s.sender.LocalUhid()
	pkt.DestinationUhid = peerUhid
	pkt.Ttl = constants.DefaultTtl
	pkt.Payload = body

	return s.sender.Send(ctx, pkt, peerUhid)
}

// Handle processes an inbound VideoCall packet: parse and fire OnCallStateChanged
// with (call_id, action, fromUhid=packet source). Returns false for the wrong
// packet type or a malformed payload (empty action counts as malformed). Returns
// an error only if the packet is nil.
func (s *Service) Handle(ctx context.Context, packet *protocol.MeshPacket) (bool, error) {
	if packet == nil {
		return false, errors.New("videocall: packet must not be nil")
	}
	if packet.Type != protocol.VideoCallPkt {
		return false, nil
	}

	var body videoCallControlWire
	if err := json.Unmarshal(packet.Payload, &body); err != nil {
		// Malformed payload: log-and-drop, not a caller error (mirrors C#).
		return false, nil
	}
	if body.Action == "" {
		return false, nil
	}

	if cb := s.OnCallStateChanged; cb != nil {
		cb(CallStateChanged{
			CallId:   body.CallID,
			Action:   body.Action,
			FromUhid: packet.SourceUhid,
		})
	}
	return true, nil
}

// videoCallControlWire is the snake_case JSON payload for PacketType.VideoCall
// call-control packets. Wire format: UTF-8 JSON, snake_case keys, field order
// call_id, action, sent_at_ms, no whitespace, lowercase-dashed UUID, sent_at_ms a
// bare integer, action an ASCII verb. This is the byte-identity gate for video
// call-control (fixtures/videocall/vectors.json). CallID is a uuid.UUID so it
// marshals to the canonical lowercase-dashed form across every language port.
// Mirrors the C# VideoCallControlPayload.
type videoCallControlWire struct {
	CallID   uuid.UUID `json:"call_id"`
	Action   string    `json:"action"`
	SentAtMs int64     `json:"sent_at_ms"`
}

File: rust/src/videocall/service.rs
// SPDX-License-Identifier: MIT

//! Directed video call-control over [`PacketType::VideoCall`].
//!
//! Ring/accept/decline/hangup signalling between two peers. The caller rings a
//! peer (minting a call id); either side then accepts, declines, or hangs up.
//! Each control verb is a directed [`PacketType::VideoCall`] send to the peer;
//! inbound signals surface via a [`VideoCallStateChangedEvent`]. The media plane
//! (SDP/ICE + frames) is handled separately by the streaming `VideoCallService`.
//! Mirrors the C# `VideoCallControlService` and the Go / Python / TS / Kotlin /
//! Swift ports.

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use tokio::sync::broadcast;
use uuid::Uuid;

use crate::constants::DEFAULT_TTL;
use crate::protocol::{MeshPacket, PacketType};
use crate::routing::sender::MeshSender;

const CALL_STATE_CHANGED_CHANNEL_CAPACITY: usize = 64;

/// JSON payload for [`PacketType::VideoCall`] packets — the video call-control
/// signal (ring / accept / decline / hangup), distinct from the media-plane
/// `VideoSignaling` (SDP/ICE) and `VideoFrame` (media) handled by the streaming
/// `VideoCallService`. This is the caller-intent layer, mirroring how `VoiceCall`
/// carries voice call-control.
///
/// Wire format: UTF-8 JSON, snake_case keys, field order `call_id`, `action`,
/// `sent_at_ms`, no whitespace, `call_id` a lowercase-dashed UUID, `sent_at_ms` a
/// bare integer, `action` an ASCII verb. Byte-identical across all eight language
/// ports — see `fixtures/videocall/vectors.json`.
#[derive(Debug, Clone, Serialize, Deserialize)]
struct VideoCallControlWire {
    /// Unique id for this call (minted by the caller on ring; echoed by accept/decline/hangup).
    call_id: Uuid,
    /// Control verb: "ring", "accept", "decline", or "hangup".
    action: String,
    /// Unix timestamp in milliseconds when the control signal was sent.
    sent_at_ms: i64,
}

/// Event emitted when a video call-control signal arrives from a peer. Mirrors the
/// C# `VideoCallStateChanged` event payload.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VideoCallStateChangedEvent {
    /// Id of the call the signal refers to.
    pub call_id: Uuid,
    /// The control verb received ("ring" / "accept" / "decline" / "hangup").
    pub action: String,
    /// UHID of the peer that sent the signal.
    pub from_uhid: String,
}

/// Directed video call-control service. Sends directed [`PacketType::VideoCall`]
/// signals (ring/accept/decline/hangup) and surfaces inbound ones via a
/// [`VideoCallStateChangedEvent`].
pub struct VideoCallControlService {
    sender: Arc<dyn MeshSender>,

    /// Broadcast channel for call-state-changed events. Each subscriber receives
    /// an event the moment an inbound [`PacketType::VideoCall`] signal is accepted.
    state_changed_tx: broadcast::Sender<VideoCallStateChangedEvent>,
}

impl VideoCallControlService {
    pub fn new(sender: Arc<dyn MeshSender>) -> Self {
        let (state_changed_tx, _) = broadcast::channel(CALL_STATE_CHANGED_CHANNEL_CAPACITY);
        Self {
            sender,
            state_changed_tx,
        }
    }

    /// Subscribe to call-state-changed events. Each subscriber receives an event
    /// the moment an inbound [`PacketType::VideoCall`] signal is accepted.
    /// Best-effort / fire-and-forget: events are dropped when there are no live
    /// receivers.
    pub fn subscribe_state_changed(&self) -> broadcast::Receiver<VideoCallStateChangedEvent> {
        self.state_changed_tx.subscribe()
    }

    /// Ring `peer_uhid`: mint a call id and send a directed "ring". Returns the new
    /// call id.
    pub async fn ring(&self, peer_uhid: &str) -> Uuid {
        let call_id = Uuid::new_v4();
        self.send_control(call_id, peer_uhid, "ring").await;
        call_id
    }

    /// Send a directed "accept" for `call_id` to `peer_uhid`. Returns delivery success.
    pub async fn accept(&self, call_id: Uuid, peer_uhid: &str) -> bool {
        self.send_control(call_id, peer_uhid, "accept").await
    }

    /// Send a directed "decline" for `call_id` to `peer_uhid`. Returns delivery success.
    pub async fn decline(&self, call_id: Uuid, peer_uhid: &str) -> bool {
        self.send_control(call_id, peer_uhid, "decline").await
    }

    /// Send a directed "hangup" for `call_id` to `peer_uhid`. Returns delivery success.
    pub async fn hangup(&self, call_id: Uuid, peer_uhid: &str) -> bool {
        self.send_control(call_id, peer_uhid, "hangup").await
    }

    async fn send_control(&self, call_id: Uuid, peer_uhid: &str, action: &str) -> bool {
        let body = serde_json::to_vec(&VideoCallControlWire {
            call_id,
            action: action.to_string(),
            sent_at_ms: unix_millis(),
        })
        .unwrap_or_default();

        let mut packet = MeshPacket::new(PacketType::VideoCall, self.sender.local_uhid());
        packet.destination_uhid = peer_uhid.to_string();
        packet.ttl = DEFAULT_TTL;
        packet.payload = body;

        self.sender.send(&packet, peer_uhid).await
    }

    /// Process an incoming [`PacketType::VideoCall`] packet: parse it and emit a
    /// [`VideoCallStateChangedEvent`]. Returns `false` for the wrong packet type, a
    /// malformed payload, or an empty action; `true` once the signal has been
    /// surfaced.
    pub async fn handle(&self, packet: &MeshPacket) -> bool {
        if packet.packet_type != PacketType::VideoCall {
            return false;
        }

        let body: VideoCallControlWire = match serde_json::from_slice(&packet.payload) {
            Ok(b) => b,
            Err(_) => return false,
        };
        if body.action.is_empty() {
            return false;
        }

        // Best-effort: deliver to any subscribers. Ignore SendError when there are
        // no live receivers (fire-and-forget).
        let _ = self.state_changed_tx.send(VideoCallStateChangedEvent {
            call_id: body.call_id,
            action: body.action,
            from_uhid: packet.source_uhid.clone(),
        });
        true
    }
}

fn unix_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

#[cfg(test)]
mod tests {
    use super::*;

    // Byte-identity gate: `VideoCallControlWire` must serialize to exactly these
    // bytes in every language (fixtures/videocall/vectors.json). snake_case keys,
    // field order call_id, action, sent_at_ms, no whitespace, UUID lowercase-dashed,
    // sent_at_ms a bare integer. Mirrors the C#
    // `VideoCallControlPayload_SerializesToCanonicalBytes` (both InlineData vectors).
    #[test]
    fn video_call_control_payload_serializes_to_canonical_bytes() {
        let cases = [
            (
                "0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f",
                "ring",
                1_700_000_000_000i64,
                "{\"call_id\":\"0f7e5d3c-1a2b-4c5d-8e9f-0a1b2c3d4e5f\",\"action\":\"ring\",\"sent_at_ms\":1700000000000}",
            ),
            (
                "00000000-0000-0000-0000-000000000000",
                "hangup",
                0i64,
                "{\"call_id\":\"00000000-0000-0000-0000-000000000000\",\"action\":\"hangup\",\"sent_at_ms\":0}",
            ),
        ];

        for (call_id, action, sent_at_ms, expected) in cases {
            let payload = VideoCallControlWire {
                call_id: Uuid::parse_str(call_id).unwrap(),
                action: action.to_string(),
                sent_at_ms,
            };
            let bytes = serde_json::to_vec(&payload).unwrap();
            let json = String::from_utf8(bytes).unwrap();
            assert_eq!(
                json, expected,
                "byte-identity mismatch for call_id={call_id} action={action}"
            );
        }
    }
}
